src/ShallowWaterRosenbrock.jl

```diff
@@ -0,0 +1,243 @@
+  RTMM = assemble_matrix(amm, U, V)
+  L2MM = assemble_matrix(amm, P, Q)
+  H1MMchol = lu(H1MM)
+  RTMMchol = lu(RTMM)
+  L2MMchol = lu(L2MM)
+
+  # Project the initial conditions onto the trial spaces
+  b₁(q)   = ∫(q*h₀)dΩ
+  b₂(v)   = ∫(v⋅u₀)dΩ
+  b₃(s)   = ∫(s*f₀)*dΩ
+  rhs1    = assemble_vector(b₃, S)
+  f       = FEFunction(S, copy(rhs1))
+  ldiv!(H1MMchol, get_free_dof_values(f))
+
+  # assemble the approximate MultiFieldFESpace Jacobian
+  n = get_normal_vector(model)
+
+  Amat((u,p),(v,q)) = ∫(-dt*λ*f*(v⋅⟂(u,n)))dΩ + ∫(dt*λ*g*(DIV(v)*p))dω - ∫(dt*λ*H₀*(q*DIV(u)))dω
+  Mmat((u,p),(v,q)) = ∫(u⋅v)dΩ + ∫(p*q)dΩ # block mass matrix
+  A = assemble_matrix(Amat, X,Y)
+  M = assemble_matrix(Mmat, X,Y)
+  B = M - A
+  Bchol = lu(B)
+  Mchol = lu(M)
+  # leap frog matrices, using 2×dt
+  lf = 1.0
+  if leap_frog
+    lf = 2.0
+  end
+  Blf = M - lf*A
+  Blfchol = lu(Blf)
+
+  # multifield initial condtions
+  b₄((v,q)) = b₁(q) + b₂(v)
+  rhs2    = assemble_vector(b₄, Y)
+  yn      = FEFunction(Y, copy(rhs2))
+  ldiv!(Mchol, get_free_dof_values(yn))
+
+  un, hn = yn
+
+  # work arrays
+  h_tmp = copy(get_free_dof_values(hn))
+  w_tmp = copy(get_free_dof_values(f))
+
+  # build the potential vorticity lhs operator once just to initialise
+  bmm(a,b) = ∫(a*hn*b)dΩ
+  H1h      = assemble_matrix(bmm, R, S)
+  H1hchol  = lu(H1h)
+
+  function run_simulation(pvd=nothing)
+    diagnostics_file = joinpath(output_dir,"nswe__rosenbrock_diagnostics.csv")
+
+    ϕ      = clone_fe_function(Q,hn)
+    F      = clone_fe_function(V,un)
+    wn     = clone_fe_function(S,f)
+    q1     = clone_fe_function(S,f)
+    q2     = clone_fe_function(S,f)
+
+    # mulifield fe functions
+    ym1     = clone_fe_function(Y,yn)
+    ym2     = clone_fe_function(Y,yn)
+    duh1    = clone_fe_function(Y,yn)
+    duh2    = clone_fe_function(Y,yn)
+    y_wrk   = copy(get_free_dof_values(yn))
+
+    if (write_diagnostics)
+      initialize_csv(diagnostics_file,"time", "mass", "vorticity", "kinetic", "potential", "power")
+    end
+
+    # first step, no leap frog
+    istep = 1
+    shallow_water_rosenbrock_time_step!(yn, ϕ, F, q1, q2, duh1, duh2, H1h, H1hchol, y_wrk,
+                                        model, dΩ, dω, Y, V, Q, R, S, f, g, ym1, ym2,
+                                        RTMMchol, L2MMchol, A, Bchol, Blfchol, dt, τ, false)
+
+    if (write_diagnostics && write_diagnostics_freq>0 && mod(istep, write_diagnostics_freq) == 0)
+      compute_diagnostic_vorticity!(wn, dΩ, S, H1MMchol, un, get_normal_vector(model))
+      dump_diagnostics_shallow_water!(h_tmp, w_tmp,
+                                      model, dΩ, dω, S, L2MM, H1MM,
+                                      hn, un, wn, ϕ, F, g, istep, dt,
+                                      diagnostics_file,
+                                      dump_diagnostics_on_screen)
+    end
+    # time step iteration loop
+    for istep in 2:N
+      get_free_dof_values(ym2) .= get_free_dof_values(ym1)
```

Comment:
I do not see why these two copies are necessary. If am not wrong, I think we can achieve the same effect using the following line, which relies on pointer association:

`ym2,ym1,yn=ym1,yn,ym2`  

BTW ... whay does m2 and m1 denote? minus 2 and minus 1?

Reply:
Yes correct, minus 1 and minus 2. I tried this previously and that is when we had the three repeating time step problem. Could be I did something else wrong but the problem was fixed by reverting to the current approach

Reply:
Yes, indeed, I was aware of that. In particular, you modified this in this commit:

https://github.com/gridapapps/GridapGeosciences.jl/pull/18/commits/0e751edc86fce4a029790b83de5985aae8fc4395

which, as you say, before you had exactly what I am proposing here. 

I have been taking a very careful look at the code, step by step, and I do no see how this line cannot work. Perhaps I can try to use the debugger to understand what might be going on.

Reply:
I was wondering if maybe the pointer swapping does not work for multifields for some reason. Probably it is something silly I did, I will double check as well...

Reply:
@davelee2804 ... FYI ... I have solved this in commit d9a0cca076e72e7d6c583f4538099071c631a568

We had two errors: 
1. If we assign `ym2` to `ym1` without storing temporarily the object `ym2` points to into an auxiliary variable, we end up assigning `yn` to `ym1` again ... (this is 1st year introduction to computer programming hehehe my fault as well)
2. We have to reassign `un,hn` at each iteration as the object `yn` points to changes at each iteration as well.

Reply:
BTW ... can 1. affect as well to the explicit solver implementation?

Reply:
Thanks for addressing thing @amartinhuertas . I would have thought 1 would indeed affect the explicit solver. I suppose it depends on the order of the pointer re-assignment. When `un`, `hn` are re-assinged in the explicit solver we don't actually need them, it is OK if they are junk when they are passed into the solver. I don't know the order of assignment in `um2,um1,un = um1,un,um2`, but if `un` is effectively the `aux` variable then I suppose we are ok...

Reply:
...Just re-checked the results of the explicit solver. There is no three-peat in the diagnostic output

Reply:
> I don't know the order of assignment in um2,um1,un = um1,un,um2

This syntax expression is equivalent to 

```
um2 = um1
um1 = un
un    =  um2
```

i.e, left to right order. 

>  I would have thought 1 would indeed affect the explicit solver. I suppose it depends on the order of the pointer re-assignment. When un, hn are re-assinged in the explicit solver we don't actually need them, it is OK if they are junk when they are passed into the solver. 

The issue is that we have pointer aliasing (if we dont use the `aux` variable trick), i.e., `um2` and `un` point to the same object. Besides, we lose track of the object that `um2` initially points to. (This would result in memory leaks in languages without GC). I think this is dirty/prone to errors, even if we have a working code. As a consequence of pointer aliasing, if `un` gets modified prior to `um1` is read, the code wont work. I guess this is not the case within the explicit time step function. (I did not check), thus explaining why it is working. I would use `aux` better to avoid potential future trouble with this.

Reply:
I've updated the explicit solver to include `u_aux`, `h_aux` in the pointer cycling in this commit here: bdeb9343d2c15479cbb52a4eb40e8b63e0afe6a5